Mount a separate filesystem for /run/setuid-wrapper-dirs

abbradar opened the pull request:
###### Motivation for this change

This makes `/run/setuid-wrapper-dirs` a separate tmpfs with `nosuid` bit unset; this makes it possible to return `nosuid` to `/run`. To ensure smooth transition, this adds more logic to activation script to check if a special filesystem already exists, and mount it rather than remount if it's not.

Tested by checking that after switch from old configuration (with `/run/setuid-wrapper-dirs` already existing, which is a more difficult case to handle) `sudo` works, and that after a clean reboot it works too.
###### Things done
- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

domenkozar commented:
Will this really work given that we call umount in the activation script?

abbradar commented:
I don't mount `/var/setuid-wrappers`, but rather `/run/setuid-wrapper-dirs` -- we don't unmount it AFAIK.

domenkozar commented:
D'oh.

domenkozar commented:
If sudo and installer tests pass with this, +1

domenkozar commented:
@abbradar huh. I'll try to reproduce.

domenkozar commented:
`machine# error: anonymous function at /tmp/root/nix/store/5macad415xi66ri7mdhf7661daif4kky-nixos-17.03.git.7ac1780/nixos/pkgs/tools/compression/gzip/default.nix:1:1 called without required argument ‘appleDerivation’, at /tmp/root/nix/store/5macad415xi66ri7mdhf7661daif4kky-nixos-17.03.git.7ac1780/nixos/lib/customisation.nix:56:12`

Something is seriously broken.

abbradar commented:
How's that:

```
machine# error: infinite recursion encountered, at undefined position
```

So random I have no idea what can be the problem...

domenkozar commented:
@abbradar on 16.09:

```
machine# finalising the installation...
machine# updating GRUB 2 menu...
machine# installing the GRUB 2 boot loader on /dev/vda...
machine# Installing for i386-pc platform.
machine# Installation finished. No error reported.
machine# setting up /etc...
machine# /nix/var/nix/profiles/system/activate: line 92: mountpoint: command not found
machine# mount: devtmpfs is already mounted or /dev busy
machine#        devtmpfs is already mounted on /dev
machine#        devtmpfs is already mounted on /tmp/root/dev
machine#        devtmpfs is already mounted on /tmp/root/mnt/dev
machine# /nix/var/nix/profiles/system/activate: line 92: mountpoint: command not found
machine# mount: devpts is already mounted or /dev/pts busy
machine#        devpts is already mounted on /dev/pts
machine#        devpts is already mounted on /tmp/root/dev/pts
machine#        devpts is already mounted on /tmp/root/mnt/dev/pts
machine# /nix/var/nix/profiles/system/activate: line 92: mountpoint: command not found
machine# /nix/var/nix/profiles/system/activate: line 92: mountpoint: command not found
machine# mount: proc is already mounted or /proc busy
machine#        proc is already mounted on /proc
machine#        proc is already mounted on /tmp/root/proc
machine#        proc is already mounted on /tmp/root/mnt/proc
machine# /nix/var/nix/profiles/system/activate: line 92: mountpoint: command not found
machine# /nix/var/nix/profiles/system/activate: line 92: mountpoint: command not found
machine# /nix/var/nix/profiles/system/activate: line 92: mountpoint: command not found
machine# mount: sysfs is already mounted or /sys busy
machine#        sysfs is already mounted on /sys
machine#        sysfs is already mounted on /tmp/root/sys
machine#        sysfs is already mounted on /tmp/root/mnt/sys
machine# Initializing machine ID from random generator.
machine: exit status 1
machine: output: 
error: command `nixos-install < /dev/null >&2' did not succeed (exit code 1)
command `nixos-install < /dev/null >&2' did not succeed (exit code 1)
```

abbradar commented:
Any news on checking? https://github.com/NixOS/nixpkgs/pull/18689/commits/ee3963775c841f5dc87b5bad462c1f5764b89a7b should have added `mountpoint` to PATH.

domenkozar commented:
Same: `machine# error: anonymous function at /tmp/root/nix/store/b3l7hhvxaw0vakyal02blzy0fh58z1lc-nixos-16.09.git.97982d4/nixos/pkgs/development/libraries/polkit/default.nix:1:1 called without required argument ‘GConf’, at /tmp/root/nix/store/b3l7hhvxaw0vakyal02blzy0fh58z1lc-nixos-16.09.git.97982d4/nixos/lib/customisation.nix:56:12`

abbradar commented:
I _think_ I remember something like that from long before. It was just like now -- random errors from different seemingly unrelated parts of nixpkgs. I only recall there was a discussion in some issue around that and it got resolved somehow, not even what triggered the error. Was too long ago and probably not helpful, sorry...

domenkozar commented:
I also remember such errors, passing arguments to the wrong functions. @nbp any ideas?

abbradar commented:
Part of this patch is superseded by f94ea048. I've rebased it; let's not include this in release (it haven't been tested and gives us only `nosuid` in `/run` which is nice but not _so_ nice to be hasty.

abbradar commented:
And now (on b69f568f4c3ebaf48a7f66b0f051d28157a61afb) it tries to download curl's sources. No idea what's going on...

abbradar commented:
So, what can we do about this? `installer.simple` does not succeed for me even on plain master.

abbradar committed:
wrappers service: make /run/wrappers a mountpoint

Also remove some compatibility code because the directory in question would be
shadowed by a mountpoint anyway.

abbradar commented:
I've rebased this atop our fresh wrappers changes. It would be nice to have this merged in 17.03 to avoid having yet potentially _more_ compatibility code :D However, I can't test this myself now because of https://github.com/NixOS/nixpkgs/issues/22868. Can anyone try to run `tests.misc.x86_64-linux` and `test.installer.simple.x86_64-linux` on this?

abbradar commented:
With fix from #23020 I can finally run NixOS tests! They both pass with this patch and I've been dogfeeding from at least September. Merging; hope this won't break much...
